### ISSUE TITLE
Could not load file or assembly UnityEngine

### ISSUE THREAD

**Issue by DawidB** [Rabadash8820/UnityAssemblies#1]
I have set up the package according to manual and I can already see UnityEngine.dll hooked up to references in the project. However when I try to create simple unit test with Debug.Log invocation, it ends with an error:
_Could not load file or assembly 'UnityEngine, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies. The system cannot find the file specified._
Did I forgot about something?

**Comment by Rabadash8820** [comment]
@DawidB Sorry for the late reply, not sure why we didn't get an email notification for this Issue...

Sometimes VS needs a little help refreshing things when you mess with project files. Have you tried closing the solution, deleting the hidden `.vs/` folder, and deleting all `bin/`, and `obj/` folders inside your project? It may "just work" when you re-open VS.

If not, could you please provide the following environment info:
- OS
- PackageReference version for this NuGet
- Unity version
- Visual Studio version
- TargetFramework

**Comment by DawidB** [comment]
@Rabadash8820 I tried again on 2019.3 and it refrenced UnityEngine DLL right away (with Intellisense recognizing it as well). However attepmting to run simple unit test wnds with file not found error: 
Unhandled Exception: System.IO.FileNotFoundException: Could not load file or 
assembly 'UnityEngine, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies. The specified module could not be found.

I have already checked path to referenced file and it's correct.

**Comment by Rabadash8820** [comment]
Hm, can you elaborate on how you're running your unit test? Are you using the Unity Test Framework, a Visual Studio test runner, or something else?

**Comment by DawidB** [comment]
I have an external solution with Xunit tests. Inside I created sample test class with simple test executing UnityEngine.Debug.Log() method.

Project settings file was updated with package reference, and following entries:
`<UnityVersion>2019.3.0f6</UnityVersion>`
`<UnityInstallRoot>D:\Program files\Unity</UnityInstallRoot>`

`<ItemGroup>`
`    <Reference Include="$(UnityInstallRoot)\$(UnityVersion)\$(UnityEnginePath)" Private="false" />`
`</ItemGroup>`

After project reload, reference to UnityEngine.dll looks fine (no exclamation in icon).

**Comment by Rabadash8820** [comment]
Hm, I'm having trouble reproducing this. I just created a .NET Standard 2.0 class library, referencing v1.3.0 of this NuGet, and v2.4.1 of Xunit, targeting Unity 2019.3.0f6. I made a simple test class like below, which is building and running without exception:

```cs
public class Class1
{
    [Fact]
    public void Test()
    {
        Debug.Log("Test");
    }
}
```

Can you tell me a little more about your environment, like the things that I mentioned in my [previous comment](https://github.com/DerploidEntertainment/UnityAssemblies/issues/1#issuecomment-570385372)?

**Comment by DawidB** [comment]
OS - Windows 10 1909
PackageReference - Unity3D.1.3.0
Unity version - 2019.3.0f6
Visual Studio version - 2019 (16.4.4)
TargetFramework - 4.6.1

I tried creating new project on .NET Core 2.1, using your unit test, but the result is the same. At least we can assume the problem is not solution-specific.

**Comment by Rabadash8820** [comment]
Crap, I think I know what's going on...  Looks like we're [using an incorrect `Condition`](https://github.com/DerploidEntertainment/UnityAssemblies/blob/5454614d60b75564c6ce66f3b5b0e45249888af8/nupkg/build/Unity3D.props#L8) in the props file, so we're just resetting whatever you set for `UnityInstallRoot` back to the default 🤦‍♂. As a temporary workaround, you should be able to set `UnityInstallPath` instead of `UnityInstallRoot` and I bet it will work. That said, I'm surprised the reference was still showing in Solution Explorer without the yellow triangle icon. Do you have multiple installations of Unity on your machine, perhaps with one still at the default path?

Also, you mentioned that you manually included a reference to `UnityEnginePath` in your project file. This is unnecessary. The MSBuild properties will pick up on the new `UnityInstallRoot` and adjust the reference accordingly, as [described here](https://github.com/DerploidEntertainment/UnityAssemblies#referencing-assemblies-at-non-default-install-locations). That is, it _will_ do that once the above bug is fixed...

**Comment by DawidB** [comment]
Looks like you're right. However I can't use UnityInstallPath as I have custom installation with non-default Unity path (C:\Program Files\Unity, not C:\Program Files\Unity\Hub\Editor), so following entry causes broken path:
`<UnityInstallPath>D:\Program files\Unity</UnityInstallPath>`

I guess I'll have to wait for a package update.

**Comment by Rabadash8820** [comment]
Yes, sorry about this, a patch version should be uploaded this evening.

For a workaround, my advice wasnt quite complete.  The current (incorrect) logic is:

1. If `UnityInstallPath` is empty, set `UnityInstallRoot` to a default value (should be if _`UnityInstallRoot`_ is empty)
1. Set all other Unity assembly paths relative to `UnityInstallRoot`

So really, your workaround would be to set `<UnityInstallRoot>D:\Program files\Unity</UnityInstallRoot>` as you were, but then also set `UnityInstallPath` to literally anything

**Comment by Rabadash8820** [comment]
@DawidB Version 1.3.1 has now been uploaded to nuget.org (may take a couple hours to show in nuget clients like the Visual Studio Package Manager). Could you confirm that `UnityInstallRoot` is working as intended after you update? Thanks!

**Comment by DawidB** [comment]
I have updated Nuget, but the issue remains (reference is correct, but unit test fails). I have set `<UnityInstallRoot>D:\Program files\Unity</UnityInstallRoot>` in project file, as well as the reference `<Reference Include="$(UnityInstallRoot)\$(UnityVersion)\$(UnityEnginePath)" Private="false" />`. This should be enough, right?

I will try linking directories to default Editor location in Unity Hub later and see if it helps.

Edit: Symbolic link does not help either.

**Comment by Rabadash8820** [comment]
Well, again, you don't need to explicitly add a `Reference` to `UnityEnginePath`, even when using a non-default `UnityInstallRoot`. That reference is added automatically, and picks up on whatever `UnityInstallRoot` you set. So you can delete that line from your csproj.

However, I doubt that that will fix your `FileNotFoundException` issue. I'm pretty stumped at this point. Do you think you could make a minimal project displaying the issue, then zip it and attach it to a comment here?

**Comment by DawidB** [comment]
You can download simple project from my [GDrive](https://drive.google.com/open?id=1_LUwPWlBpIGmV3OJCuqN1C4tMk38yX68).

**Comment by Rabadash8820** [comment]
@DawidB Thanks, I'll take a look at that and see if I can figure out the problem.

**Comment by Rabadash8820** [comment]
@DawidB Ah, okay finally I was able to reproduce your issue using that project. Looks like the issue was in your `TargetFramework`. Unity only supports targeting Mono (.NET 4.x equivalent) or .NET Standard 2.0. You were targeting .NET Core 2.1, so the CoreCLR was trying to load an assembly named `UnityEngine` compatible with that runtime, but it couldn't find one, hence the `FileNotFoundException`. Changing the `TargetFramework` to `netstandard2.0` allows the test to run successfully.

Also, you do not need to redefine the `UnityInstallRoot` property if you are defining `OSInstallRoot`. The former will automatically use the value of the latter. We should probably update our Readme to clarify all this, and mention the lack of .NET Core support. On a side note, the `xunit.runner.visualstudio` package reference showed a warning when I switched to .NET Standard. You should be able to just delete that package and still run tests fine in Visual Studio.

Long story short, here is the minimal project file that should work for you (and don't forget that `H:\` is a Windows-only path so this project will not build on Mac or Linux 😉):

**Edit: Forgot to add `Program Files` to the `OSInstallRoot`**

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <UnityVersion>2019.3.0f6</UnityVersion>
    <OSInstallRoot>H:\Program Files</OSInstallRoot>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Unity3D" Version="1.3.1" />
    <PackageReference Include="xunit" Version="2.4.1" />
  </ItemGroup>

</Project>
```

**Comment by DawidB** [comment]
@Rabadash8820 Sorry for creating confusion with .NET Core target. I have forgotten it's not supported by Unity.

I tried reverting it back to .NET Standard 2.0, but still did not managed to fix the problem. First of all - settings `<OSInstallRoot>D:\</OSInstallRoot>` alone does not seem to work on my environment, as assemblies list still show reference to default `C:\Program files...` location. However it can be fixed by adding one more line: `<UnityInstallRoot>$(OSInstallRoot)\Program files\Unity</UnityInstallRoot>`
This fixes missing reference problem, but I cannot run test on .NET standard (nothing happens when I try to, i guess visualstudiotestrunner nuget is required after all).

I tried to create a project with .NET Framework 4.6 as target, but the final result is a test with FileNotFoundException result.

[Link to both projects](https://drive.google.com/file/d/1zmC7tYhpUDrLjEUJ3CSEV-xqa3cqhEGY/view?usp=sharing)

**Comment by Rabadash8820** [comment]
Hm, you might be right about the test runner NuGet. I don't really use XUnit so I'm not sure which packages are required. As long as you restore it using the correct target framework you should be okay; the warning that I mentioned was just something about having restored it using .NET Core in a .NET Standard project.

My bad on the `OSInstallRoot` thing, I mistyped in my minimal csproj above (see edit). The actual value of that property should be `H:\Program Files`. Now, that's assuming that the DLL path you ultimately want to get to is `H:\Program Files\Unity\Hub\Editor\$(UnityVersion)\Editor\Data\Managed\UnityEngine.dll`. It looks like you might not be using Unity Hub, and instead want to get to the path `H:\Program Files\Unity\$(UnityVersion)\Editor\Data\Managed\UnityEngine.dll`. In that case, you can just skip the `OSInstallRoot` and define `<UnityInstallRoot>H:\Program Files\Unity</UnityInstallRoot>`.

As for .NET 4.6, I'm not sure why you'd be having issues there. I'll take a look at your sample project this evening.

**Comment by Rabadash8820** [comment]
@DawidB I am also seeing errors when trying to build your .NET Framework 4.6 project. Strangely, I was getting `SecurityException`s, not `FileNotFoundException`s, until I tweaked your `packages.config` file to use `targetFramework="net46"` instead of `net472` for all packages, then I started getting `FileNotFoundException`. In both cases, I couldn't get the project to actually build until I unloaded and reloaded it once. I'd be interested to see how your errors would change if you used an SDK-style net46 project, rather than the older MSBuild Tools style project that you're using currently.

All that said, it looks like the issue you're running into is that you are trying to execute Unity's core APIs (those implemented in the native, C++-based engine) from a .NET runtime, i.e., from the Visual Studio test runner. That is not going to work, as has [been](https://github.com/JetBrains/resharper-unity/issues/96#issuecomment-270001428) [noted](https://forum.unity.com/threads/c-error-ecall-methods-must-be-packaged-into-a-system-module.199361/) [by](https://forum.unity.com/threads/security-exception-ecall-methods-must-be-packaged-into-a-system-module.98230/) [others](https://stackoverflow.com/questions/30737729/c-sharp-security-exception) before. I probably should've caught that you were trying to do this before. This NuGet allows you to _reference_ Unity's APIs in netstandard2.0 and net46 (and above?) projects so that the project's will _build_, but you will never be able to _run_ those projects outside of Unity. So instead of making executable .NET Framework projects, I highly recommend you make .NET Standard _Class Library_ projects, which can then be imported into Unity as [managed plugins](https://docs.unity3d.com/Manual/UsingDLL.html). For running unit tests, specifically, you should also take a look at the [Unity Test Framework](https://docs.unity3d.com/Packages/com.unity.test-framework@1.1/manual/index.html) package, which replaces the older Unity Test Runner system in 2019.2+.

I realize that our Readme never explicitly says that projects using this NuGet cannot be run directly. That's something we'll need to clarify.

**Comment by DawidB** [comment]
@Rabadash8820 Thanks for detailed explanation. I am actually using Unity Test Framework in my project for most of the test. The reason for my attempt to use this Nuget was a separate solution that I have in my project. It references some classes from main project and it contains Xunit unit tests. Thanks to such setup I was able to quickly code and run tests without waiting for Unity compilation. So in the end it looks like I will have to fall back to Unity Test Framework with those separate tests, which is not that bad, as maintaining separate solution is cumbersome.

**Comment by Rabadash8820** [comment]
@DawidB No problem :) Sorry for the anti-climactic solution after all this back and forth. I'm going to go ahead and close this Issue. If something else comes up, don't hesitate to open a new Issue.

**Comment by Rabadash8820** [comment]
FWIW, it is still possible to have multiple solutions and/or multiple projects when working on a Unity project, you just have to be careful of what references what. For example, you could have a setup like the following (this is what we do for Derploid games; I'd share a link but those repos are private, unfortunately):

- `MyGame.dll`: .NET Standard class library for miscellaneous types calling Unity APIs, imported into Unity as a managed plugin via post-build targets, which references:
- `MyGame.*.dll`: zero or several .NET Standard class libraries for game systems (e.g., `MyGame.Inventory.dll`) that _do not_ call Unity APIs, also imported into Unity as managed plugins. Referenced by:
- `MyGame.*.Test.dll`: zero or several .NET Core test projects that test the types defined in the corresponding .NET Standard class library. We prefer NUnit and Moq, but could obviously have used XUnit, MSTest, or similar. The `MyGame.*.dll` class libraries expose internals to these projects for whitebox(ish) testing. The `Test` DLLs are _not_ imported into Unity.
- `MyGame.Test.EditMode`: a .NET Standard assembly that provides [edit mode tests](https://docs.unity3d.com/Packages/com.unity.test-framework@1.1/manual/edit-mode-vs-play-mode-tests.html#edit-mode-tests) for `MyGame.dll`. The source code for this assembly is actually stored in an Editor/ folder somewhere under Assets/, with a corresponding .asmdef file. To code in Visual Studio, we have an empty project that includes the source files through:

```xml
<ItemGroup>
    <Compile Include="/path/to/Assets/subpath/Editor/Test.EditMode/**/*.cs" />
</ItemGroup>
```
- `MyGame.Test.PlayMode`: a .NET Standard assembly that provides [play mode tests](https://docs.unity3d.com/Packages/com.unity.test-framework@1.1/manual/edit-mode-vs-play-mode-tests.html#play-mode-tests) for `MyGame.dll`. The source code for this assembly is stored somewhere under Assets/ (_not_ under an Editor/ folder), with a corresponding .asmdef file. To code in Visual Studio, we again have an empty project that includes the source files through `Compile` items like above.
- The `EditMode` and `PlayMode` projects target .NET Standard because of the Unity project settings, not from any .csproj file. Only these tests require the Unity Test Framework runner. All of the `MyGame.*.Test.dll` assemblies and the game assemblies that they reference can be built/run completely outside of Unity. Its unclear whether the edit/play mode test source files have to be kept under Assets/ with an .asmdef file for Unity Test Framework to recognize them, but we had issues trying to import such tests from managed plugins in the past, so we now stick to this method.
